js/eventloop/eventloop.go

```diff
@@ -55,11 +55,69 @@ func (e *EventLoop) wakeup() {
+// some asynchronous work off the main thread and that you may need to execute
+// some code back on the main thread when you are done. So, once you call this
+// method, the event loop will wait for you to finish and give it the callback
+// it needs to run back on the main thread before it can end the whole current
+// script iteration.
+//
+// ReservePendingCallback() *must* be called from the main runtime thread, but
+// its result scheduleOnMainThread() is thread-safe and can be called from any
+// goroutine. scheduleOnMainThread() ensures that its callback parameter is
+// added to the VU runtime's tasks queue, to be executed on the main runtime
+// thread eventually, when the VU is done with the other tasks before it. Unless
+// the whole event loop has been stopped, invoking scheduleOnMainThread() will
+// queue its argument and "wake up" the loop (if it was idle, but not stopped).
+//
+// Keep in mind that once you call ReservePendingCallback(), you *must* also
+// call scheduleOnMainThread() exactly once, even if don't actually need to run
+// any code on the main thread. If that's the case, you can pass an empty no-op
+// callback to it, but you must call it! The event loop will wait for the
+// scheduleOnMainThread() invocation and the k6 iteration won't finish and will
+// be stuck until the VU itself has been stopped (e.g. because the whole test or
+// scenario has ended). Any error returned by any callback on the main thread
+// will abort the current iteration and no further event loop callbacks will be
+// executed this iteration.
+//
+// A common pattern for async work is something like this:
+//
+//    func doAsyncWork(vu modules.VU) *goja.Promise {
+//        scheduleOnMainThread := vu.ReservePendingCallback()
+//        p, resolve, reject := vu.Runtime().NewPromise()
+//
+//        // Do the actual async work in a new independent goroutine, but make
+//        // sure that the Promise resolution is done on the main thread:
+//        go func() {
+//            // Also make sure to abort early if the context is cancelled, so
+//            // the VU is not stuck when the scenario ends or Ctrl+C is used:
+//            result, err := doTheActualAsyncWork(vu.Context())
+//            scheduleOnMainThread(func() error {
+//                if err != nil {
+//                    reject(err)
+//                } else {
+//                    resolve(result)
+//                }
+//                return nil  // do not abort the iteration
+//            })
+//        }()
+//
+//        return p
+//    }
+//
+// This ensures that the actual work happens asynchronously, while the Promise
+// is immediately returned and the main thread resumes execution. It also
+// ensures that the Promise resolution happens safely back on the main thread
+// once the async work is done, as required by goja and all other JS runtimes.
+func (e *EventLoop) ReservePendingCallback() (scheduleOnMainThread func(callback func() error)) {
```

Comment:
And the alternative, `asyncslot.RunOnMain(func() error {` or `asyncslot.RunTask(func() error{`

Reply:
Hmm, separate types are probably going to help with clarity a bit, but I don't want to grow the scope of this beyond a rename right now... :thinking: maybe something to consider in https://github.com/grafana/k6/issues/2544?

> And the relative alternatives, `asyncslot.RunOnMain(func() error {` or `asyncslot.RunTask(func() error{`

I see a couple of problems in this suggestion:
- `asyncslot` is wrong because the slot we are reserving is actually for synchronous work :sweat_smile: we are doing the asynchronous work in a separate thread (or 10, or 100) ourselves, and then we are reserving a slot for synchronous work on the main thread so we can call some callbacks or resolve promises or whatever
- `RunOnMain` and `RunTask` are misleading because they won't actually run whatever callback you give them immediately, they will schedule it to run in the future by adding it to the queue of pending task to be executed on the main thread

Reply:
>Hmm, separate types are probably going to help with clarity a bit, but I don't want to grow the scope of this beyond a rename right now... thinking maybe something to consider in https://github.com/grafana/k6/issues/2544?

IMHO, that would be the way how we should proceed, but we are deprecating+adding methods here, so adding now a method for eventually breaking it again in the future doesn't sound good to me. So, should we just document here and discuss all the rest in https://github.com/grafana/k6/issues/2544 or in the dedicated PR?

Reply:
I doubt we'll ever remove `RegisterCallback` completely, `TaskQueue` depends on it after all :thinking: Right, @mstoykov? So we maybe can un-export it in the future, but it's still worth figuring out a non-confusing name for it.

Reply:
I guess we can remove it entirely, but I doubt we will ever do it. So I am also for at least documenting it better. Renaming it again is dependent on people agreeing it's a big enough of an improvement to warrant us going through the trouble of deprecating and rewriting everything that uses it :( 

I very much doubt we will remove it though as it's a pretty good building block which is probably how any of the future "porcelain" on top will need to be implemented either way. 

Having the callbacks typed requires that we have a type somewhere and as this is uses primarily from the `js/modules` sub-package - it makes sense to be there ... but on the other hand it makes sense to be in the `js/eventloop`. I did not go this way as I didn't want to import two packages everywhere to use the API and at least at the time to me it didn't really seem to improve the situation all that much :shrug: . Again this is probably better decided by not the person who has designed this and knows how it works ;).